concordium-smart-contract-testing/src/lib.rs

```diff
@@ -0,0 +1,2745 @@
+                                } else {
+                                    // Move the CCD
+                                    self.chain
+                                        .get_account_mut(to)
+                                        .expect("Account is known to exist")
+                                        .balance += amount;
+                                    let instance = self
+                                        .chain
+                                        .get_instance_mut(self.address)
+                                        .expect("Contract is known to exist");
+                                    instance.self_balance -= amount;
+
+                                    // Add transfer event
+                                    self.chain_events.push(ChainEvent::Transferred {
+                                        from: self.address,
+                                        amount,
+                                        to,
+                                    });
+
+                                    InvokeResponse::Success {
+                                        new_balance: instance.self_balance,
+                                        data:        None,
+                                    }
+                                }
+                            };
+                            let success = matches!(response, InvokeResponse::Success { .. });
+                            // Add resume event
+                            self.chain_events.push(ChainEvent::Resumed {
+                                address: self.address,
+                                success,
+                            });
+
+                            let resume_res = v1::resume_receive(
+                                config,
+                                response,
+                                InterpreterEnergy::from(remaining_energy),
+                                &mut self.mutable_state,
+                                false, // never changes on transfers
+                                self.loader,
+                            );
+
+                            // Resume
+                            self.process(resume_res)
+                        }
+                        v1::Interrupt::Call {
+                            address,
+                            parameter,
+                            name,
+                            amount,
+                        } => {
+                            // Add the interrupt event
+                            self.chain_events.push(interrupt_event);
+
+                            if state_changed {
+                                println!("\t\tState was changed. Saving prior to another call.");
+                                let mut collector = SizeCollector::default();
+                                let persistent_state =
+                                    self.mutable_state.freeze(&mut self.loader, &mut collector);
```

Comment:
I don't think this is the way to go. The mutable state itself already has support for checkpointing and rollbacks. Freezing is expensive and is not something that happens on interrupts. It only happens at the very end of execution before transaction commit. So in reference to TODO below, no, we don't charge here.